### DISCUSSION TITLE
Redirect user when Stripe webhook stuff is done

### DISCUSSION THREAD

**Discussion by GuysmoB** [vercel/next.js#66410] [Help]
### Summary

Hi,

I have a Stripe webhook triggered when someone makes a checkout on my app.
When checkout is done, the user is redirected to another page while the webhook is triggered.
Mutation data is stored inside Stripe checkout metadata so I can access it inside the webhook.
Mutations are done, but at this moment, the user has already been redirected to the page where data will change.
However, the redirect is faster than the mutation from the webhook, so data aren't up to date, even with a revalidatePath used at the end of the mutation.

1 - Do you have any ideas on how to better handle this flow?
2 - Is it possible to wait for the webhook to finish before redirecting the user?

Thanks for any hints !

**Comment by adopstarliu** [comment]
I have the same question and my solution is to write client side logic in payment.paymentIntent.status case 'succeeded', to pool an api route or server action to check if the database insertion has finished, per second or something like that and once insertion finishes, then redirect. 

It's been a couple of months, have you found a solution?

**Reply by GuysmoB** [reply]
I did exactly like you.
I created my object in database before the payment with a boolean isPaymentSuccessfull= false
I created a custom form to with a paymentIntent.  
If payment is succesfull, i update my object with isPaymentSuccessfull= true so it's visible for the user only after payment and i can handle any redirect or mutations. I no longer use the webhook.

I did that to prevent any loss of informations if something happen client side.

**Comment by Ilia01** [comment]
I would not try to make the redirect wait for the webhook.

Stripe's redirect and Stripe's webhook are two separate asynchronous channels, and treating the webhook as something the browser should block on usually makes both the UX and the architecture worse.

The usual pattern is:

1. complete checkout
2. redirect the user immediately to a success or processing page
3. let the webhook idempotently write the final state
4. have the client poll, revalidate, or subscribe until that state shows up

If you need something synchronous for the redirect itself, put that state change in the server action or API call that created the Checkout Session, not in the webhook. The webhook should be the authoritative confirmation path, not the user's navigation gate.

`revalidatePath()` can help with cache invalidation, but it does not change the fact that the webhook finishes after the redirect.

So the short answer is:

- no, I would not wait for the webhook before redirecting
- yes, I would render a pending state and refresh from canonical server state until the webhook-backed mutation lands
